### PR TITLE
Undo cost tracker changes for retryable tx

### LINKED ISSUE #23030: QoS Service Over Estimates Compute Units
#### Problem
The QoS Service selects + updates the cost model before building a batch, resulting in an overestimate of compute units: https://github.com/solana-labs/solana/blob/f67a27eee/core/src/banking_stage.rs#L1119

discord:
https://discord.com/channels/428295358100013066/754723351389536348/940847277134729266

#### Proposed Solution
- make cost model + compute units accurate.
- can take into account the batch that's actually built + executed.
- probably want to take an iterative approach to cost model while building the batch? maybe you want to run qos while executing the tx?

cc @jstarry

### PR DESCRIPTION
**Pull request by ckamm**
#### Problem

#23030. The cost tracker values get adjusted for transactions that don't get executed and may be retried.

#### Summary of Changes

Previously tx that need to be retried (like due to AccountInUse) affected the CostTracker. Meaning that their cost would be added again when the tx was retried.

Now, retryable accounts have their costs removed from the CostTracker again.

This is a draft because I still want to measure the performance impact and add a test.

I also tried to make the patch easy to backport. It may be worthwhile to adjust the way the locking and cost model interact entirely. Or even use the real-cu in the cost tracker instead of estimated-cu.

Fixes #23030

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
@taozhu-chicago

**Comment by taozhu-chicago** [comment]
Thank you @ckamm for identifying this. It is a bug that contributes the low density blocks. 

For context: 
- one cost_tracker per block, shared between banking_threads (via qos_service)
- cost_tracking is a `check then commit` two-step op. 

Earlier it was implemented as: would a tx with `X` estimated units fit (Check)? if so, execute tx; and if successful, commit `X` units to cost tracker, otherwise no change to cost_tracker. So retried TXs are not counted in cost_tracker. However, the issues with this are: 
- requires additional RwLocking and lookup, has noticeable performance penalty
- more importantly, since cost_tracker is shared amount banking_threads, committing cost after execution causes block over-packing, which is something we want to avoid. For example, cost_tracker currently reads 1_000 units have used. Thread_1 and thread_2 both check if they can fit a 10_000 units. If the limit is 11_000, then the answers are yes and yes, allowing both threads to execute their 10_000 units tx, resulting a block of (1_000 + 10_000 + 10_000 > 11_000). 

To prevent over-packing, the cost_tracker commit part was move _before_ tx execution, so leader is more conservative, the price paid is block not being packed at its fullness. 

Changes in this PR would address the issue, if performance overhead is acceptable. (can use `banking_bench` or `1-hour perf test` to measure) 

Like the idea of counting actual units to cost_tracker, possibly to book cost_tracker with estimated-cu, then commit with actual-cu after execution.

**Comment by taozhu-chicago** [comment]
Potentially this approach #24182 ? It includes plumbing work for using transaction's actual unit when commit to cost_tracker.

File: core/src/qos_service.rs [hunk 1]
```diff
@@ -170,6 +170,33 @@ impl QosService {
...
+        transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
+        retryable_transaction_indexes: &[usize],
+        bank: &Arc<Bank>,
+    ) {
+        let mut cost_tracker = bank.write_cost_tracker().unwrap();
+        transaction_costs
+            .zip(transaction_qos_results)
+            .enumerate()
+            .for_each(|(index, (tx_cost, qos_result))| {
+                if qos_result.is_ok() && retryable_transaction_indexes.contains(&index) {
```
Comment (ckamm) [inline core/src/qos_service.rs:188]: jstary pointed out that a tx can be unexecuted without ending up in retryable_transaction_indices. Needs a fix.

File: core/src/banking_stage.rs [hunk 2]
```diff
@@ -1351,38 +1351,40 @@ impl BankingStage {
...
         qos_service: &QosService,
     ) -> ProcessTransactionBatchOutput {
-        let ((transactions_qos_results, cost_model_throttled_transactions_count), cost_model_time) =
-            Measure::this(
-                |_| {
-                    let tx_costs = qos_service.compute_transaction_costs(txs.iter());
+        let (
+            (transactions_qos_results, cost_model_throttled_transactions_count, transaction_costs),
+            cost_model_time,
+        ) = Measure::this(
```
Comment (taozhu-chicago) [inline core/src/banking_stage.rs:1357]: nit: `Measure::this` doesn't really help to make code cleaner when it has to pass out many results. May as well just call `Measure::start` and `stop`.

File: core/src/banking_stage.rs [hunk 3]
```diff
@@ -1397,21 +1399,28 @@ impl BankingStage {
...
         let ExecuteAndCommitTransactionsOutput {
             ref mut retryable_transaction_indexes,
             ref execute_and_commit_timings,
             ..
         } = execute_and_commit_transactions_output;
 
+        QosService::update_or_remove_transaction_costs(
+            transaction_costs.iter(),
+            transactions_qos_results.iter(),
+            retryable_transaction_indexes,
```
Comment (taozhu-chicago) [inline core/src/banking_stage.rs:1416]: @jstarry 's [comment](https://github.com/solana-labs/solana/pull/24182#discussion_r846204232) should be address. I'd suggested create issue to work on separately, as that is not as pressing as those retryable failures. wdyt?

File: core/src/banking_stage.rs [hunk 4]
```diff
@@ -1397,21 +1399,28 @@ impl BankingStage {
...
+        drop(batch);
+        unlock_time.stop();
+
         let ExecuteAndCommitTransactionsOutput {
             ref mut retryable_transaction_indexes,
             ref execute_and_commit_timings,
             ..
         } = execute_and_commit_transactions_output;
 
+        QosService::update_or_remove_transaction_costs(
```
Comment (taozhu-chicago) [inline core/src/banking_stage.rs:1413]: +1 to move it to qos_service

File: runtime/src/cost_model.rs [hunk 5]
```diff
@@ -16,14 +16,15 @@ use {
 const MAX_WRITABLE_ACCOUNTS: usize = 256;
 
 // costs are stored in number of 'compute unit's
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (taozhu-chicago) [inline runtime/src/cost_model.rs:19]: maybe `Clone` is not necessary after all

**Review by taozhu-chicago — APPROVED** [review]
lgtm, +nits.

File: core/src/banking_stage.rs (hunk 3 above)
Comment (taozhu-chicago) [reply core/src/banking_stage.rs:1416]: added issue #24197

Comment (ckamm) [reply core/src/banking_stage.rs:1416]: I agree. The current state is already an improvement, even if a part of the problem still remains.

**Review by taozhu-chicago — APPROVED** [review]
lgtm

File: core/src/banking_stage.rs [hunk 6]
```diff
@@ -1435,6 +1442,30 @@ impl BankingStage {
...
+        transaction_results: impl Iterator<Item = &'a transaction::Result<()>>,
+        retryable_transaction_indexes: &[usize],
+        qos_service: &QosService,
+        bank: &Arc<Bank>,
+    ) {
+        transactions
+            .zip(transaction_results)
+            .enumerate()
+            .for_each(|(index, (tx, result))| {
+                if result.is_ok() && retryable_transaction_indexes.contains(&index) {
```
Comment (t-nelson) [inline core/src/banking_stage.rs]: if we pass in `transactions` and `transaction_resutls` as slices and instead iterate on `retryable_transaction_indexes` here, we can safely index into the slices and avoid the `.contains()` O(N) lookup

File: runtime/src/cost_tracker.rs [hunk 7]
```diff
@@ -89,51 +98,33 @@ impl CostTracker {
...
-        self.would_fit(
-            &tx_cost.writable_accounts,
-            tx_cost.sum(),
-            tx_cost.account_data_size,
-            transaction,
-        )
+    ) -> Result<u64, CostTrackerError> {
+        self.would_fit(tx_cost)?;
+        self.pending_transactions
+            .insert(*transaction.signature(), tx_cost.clone());
```
Comment (t-nelson) [inline runtime/src/cost_tracker.rs]: It would be nice to avoid this clone.  `TransactionCost`s are large

Comment (ckamm) [reply runtime/src/cost_tracker.rs]: This has already been resolved, the clone is gone

File: core/src/banking_stage.rs (hunk 6 above)
Comment (ckamm) [reply core/src/banking_stage.rs]: Agreed. @taozhu-chicago is fixing a bug in this area anyway, and we'll likely iterate over `transaction_qos_results.zip(transaction_execution_results)` or something, and won't need the retryable_indices at all.

**Pull request merged**